### ISSUE TITLE
Attempting to update an existing alert results in it being deleted from the database

### ISSUE THREAD

**Issue by nfm** [metabase/metabase#12350]
**Describe the bug**
After creating an email alert, when I try to modify it (to add an additional recipient), the alert is deleted entirely.

**Logs**
Happy to pull and sanitize these if needed - please let me know.

**To Reproduce**
1. Create an email alert on a question
2. Refresh the question page
3. Click the blue alert bell, then 'Edit'
4. Add an additional email recipient
5. Click 'Save changes'

The client first issues a PUT request to `/api/card/:id`, which appears to result in the existing alert being deleted. I receive a "One of your alerts has stopped working" email to that effect.

Then the client attempts to issue a PUT request to `/api/alert/:id`, which 404s, because the alert has been deleted.

I can successfully retrieve the alert from the database after step 1. above. It's then gone after step 5.

**Expected behavior**
Alert is updated, and not deleted.

**Screenshots**
n/a

**Information about your Metabase Installation:**

```json
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:75.0) Gecko/20100101 Firefox/75.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_242-heroku-b08",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_242-heroku",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.242-b08",
    "os.name": "Linux",
    "os.version": "4.4.0-1062-aws",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "heroku",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "9.6.17"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.35.2",
      "date": "2020-04-10",
      "branch": "release-0.35.x",
      "hash": "f3b2857"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}
```

**Severity**
Moderate. In conjunction with #12349, this means it's not currently possible to create alerts that are Slack only.

**Additional context**
n/a

**Comment by flamber** [comment]
Hi @nfm 
I'm not able to reproduce this issue. Can you supply the parameters+response from the requests? Redact sensitive (emails) info as needed.

**Comment by nfm** [comment]
Thanks @flamber. There's definitely something odd going on here, I'm not sure why submitting the form to update an alert is resulting in a request to update the question.

After I hit 'Save changes' on the alert, these three requests occur (in this order):

**1. PUT /api/card/490 [202 Accepted]**

<details>
<summary>Request params</summary>

```json

{"name":"Annual subscriptions created per day by state","dataset_query":{"database":3,"type":"query","query":{"source-table":68,"filter":["and",["=",["fk->",["field-id",841],["field-id",911]],"annual"],["time-interval",["field-id",829],-45,"day",{"include-current":true}],["!=",["field-id",847],"errored"]],"aggregation":[["count"]],"breakout":[["datetime-field",["field-id",829],"day"],["field-id",851]]}},"display":"bar","description":null,"visualization_settings":{"graph.dimensions":["created_at","stripe_status"],"graph.metrics":["count"],"graph.show_values":false},"archived":false,"enable_embedding":false,"embedding_params":null,"collection_id":104,"collection_position":null,"result_metadata":[{"display_name":"Created At","name":"created_at","base_type":"type/Date","special_type":"type/CreationTimestamp","unit":"day","fingerprint":{"global":{"distinct-count":2980,"nil%":0},"type":{"type/DateTime":{"earliest":"2015-10-15T01:57:45.033Z","latest":"2019-04-08T18:13:56.619Z"}}}},{"display_name":"Stripe Status","name":"stripe_status","base_type":"type/PostgresEnum","special_type":"type/Category","fingerprint":{"global":{"distinct-count":6,"nil%":0.1681208053691275},"type":{"type/Text":{"percent-json":0,"percent-url":0,"percent-email":0,"average-length":5.11006711409396}}}},{"display_name":"Count","name":"count","base_type":"type/BigInteger","special_type":"type/Quantity","fingerprint":{"global":{"distinct-count":9,"nil%":0},"type":{"type/Number":{"min":1,"max":24,"avg":3.9722222222222223,"sd":4.27942660452216,"q1":1.4274007043062173,"q3":5}}}}],"metadata_checksum":"Www0FCyoaEhH5K/x0idfqg=="}
```

</details>

<details>
<summary>Response</summary>

```json
{"description":null,"archived":false,"collection_position":null,"table_id":68,"result_metadata":[{"display_name":"Created At","name":"created_at","base_type":"type/Date","special_type":"type/CreationTimestamp","unit":"day","fingerprint":{"global":{"distinct-count":2980,"nil%":0},"type":{"type/DateTime":{"earliest":"2015-10-15T01:57:45.033Z","latest":"2019-04-08T18:13:56.619Z"}}}},{"display_name":"Stripe Status","name":"stripe_status","base_type":"type/PostgresEnum","special_type":"type/Category","fingerprint":{"global":{"distinct-count":6,"nil%":0.1681208053691275},"type":{"type/Text":{"percent-json":0,"percent-url":0,"percent-email":0,"average-length":5.11006711409396}}}},{"display_name":"Count","name":"count","base_type":"type/BigInteger","special_type":"type/Quantity","fingerprint":{"global":{"distinct-count":9,"nil%":0},"type":{"type/Number":{"min":1,"max":24,"avg":3.9722222222222223,"sd":4.27942660452216,"q1":1.4274007043062173,"q3":5}}}}],"creator":{"email":"redacted@redacted.com","first_name":"Redacted","last_login":"2020-04-14T05:24:47.622904Z","is_qbnewb":false,"is_superuser":true,"id":3,"last_name":"Redacted","date_joined":"2017-03-28T23:02:19.972Z","common_name":"Redacted"},"can_write":true,"database_id":3,"enable_embedding":false,"collection_id":104,"query_type":"query","name":"Annual subscriptions created per day by state","dashboard_count":0,"read_permissions":null,"creator_id":3,"updated_at":"2020-04-22T04:54:14.424332Z","made_public_by_id":null,"embedding_params":null,"cache_ttl":null,"dataset_query":{"database":3,"type":"query","query":{"source-table":68,"filter":["and",["=",["fk->",["field-id",841],["field-id",911]],"annual"],["time-interval",["field-id",829],-45,"day",{"include-current":true}],["!=",["field-id",847],"errored"]],"aggregation":[["count"]],"breakout":[["datetime-field",["field-id",829],"day"],["field-id",851]]}},"id":490,"display":"bar","visualization_settings":{"graph.dimensions":["created_at","stripe_status"],"graph.metrics":["count"],"graph.show_values":false},"collection":{"id":104,"name":"New pricing 2020","description":null,"color":"#509EE3","archived":false,"location":"/","personal_owner_id":null,"slug":"new_pricing_2020"},"created_at":"2020-03-24T04:21:25.131Z","public_uuid":null}
```

</details>

**2. GET /api/alert/question/490 [200 Ok]**

Request params: (none)
Response: `[]`

**2. PUT /api/alert/26 [404 Not Found]**

<details>
<summary>Request params</summary>

```json

{"created_at":"2020-04-20T01:31:50.996337Z","skip_if_empty":true,"alert_first_only":false,"updated_at":"2020-04-20T01:31:50.996337Z","card":{"id":490,"include_csv":true,"include_xls":false},"creator_id":1,"name":null,"alert_condition":"rows","channels":[{"created_at":"2020-04-20T01:31:50.996337Z","enabled":false,"schedule_day":null,"id":50,"pulse_id":26,"updated_at":"2020-04-20T01:31:50.996337Z","recipients":[{"id":1,"email":"redacted@redacted.com","first_name":"Redacted","last_name":"Redacted","common_name":"Redacted"}],"schedule_frame":null,"channel_type":"email","schedule_hour":0,"schedule_type":"daily"},{"channel_type":"slack","enabled":true,"recipients":[],"details":{},"schedule_type":"daily","schedule_day":null,"schedule_hour":0,"schedule_frame":null}],"can_write":true,"creator":{"email":"redacted@redacted.com","first_name":"Redacted","last_login":"2020-04-08T23:47:25.126065Z","is_qbnewb":false,"is_superuser":true,"id":1,"last_name":"Redacted","date_joined":"2017-03-14T05:06:25.613Z","common_name":"Redacted"},"collection_position":null,"archived":false}

```

</details>

In this instance, I verified that the metabase database had an alert with ID=26 persisted prior to submission.

I don't think this happens on _all_ questions.

Does saving a question have an intended side-effect of deleting alerts associated with it?

If so, maybe the bug is just that the question is being saved when it needn't be.

Let me know if there's anything else it'd be helpful to provide, or if there's a better format to share the request/response details in :grin:

**Comment by flamber** [comment]
@nfm 
Okay, I've tried multiple times, but I don't get the error on `PUT /api/alert/:id` - it's always `200`.

I followed your steps-to-reproduce exactly, so not sure why you're getting an error, nor why it's 404, since you're updating an existing alert. I don't see anything in the code that should enable alerts to be deleted (or added or updated) with `PUT /api/card/:id`.
Alerts are deleted with `DELETE /api/alert/:id`, added with `POST /api/alert` and updated with `PUT /api/alert/:id`

I agree that it seems odd that it's saving the question first, which also causes an empty revision history as well. I'm unsure if it was added to "track" alert changes. Related issues: #1926 and #3801
